VideoWeb/VideoWeb/ClientApp/src/app/shared/mic-visualiser/mic-visualiser.component.spec.ts

```diff
@@ -0,0 +1,62 @@
+import { ChangeDetectorRef } from '@angular/core';
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+import { By } from '@angular/platform-browser';
+
+import { MicVisualiserComponent } from './mic-visualiser.component';
+
+describe('MicVisualiserComponent', () => {
+    let component: MicVisualiserComponent;
+    let fixture: ComponentFixture<MicVisualiserComponent>;
+    let changeDetectorSpy: jasmine.SpyObj<ChangeDetectorRef>;
+    let canvasElement;
+    let canvasContext;
+
+    beforeEach(async () => {
+        changeDetectorSpy = jasmine.createSpyObj<ChangeDetectorRef>('ChangeDetectorRef', ['detectChanges']);
+
+        await TestBed.configureTestingModule({
+            declarations: [MicVisualiserComponent],
+            providers: [{ provide: ChangeDetectorRef, useValue: changeDetectorSpy }]
+        }).compileComponents();
+    });
+
+    beforeEach(() => {
+        fixture = TestBed.createComponent(MicVisualiserComponent);
+        component = fixture.componentInstance;
+        fixture.detectChanges();
+
+        canvasElement = fixture.debugElement.query(By.css('#meter')).nativeElement;
+        canvasContext = canvasElement.getContext('2d');
+    });
+
+    it('should create', () => {
+        expect(component).toBeTruthy();
+    });
+
+    it('should call setupStream when there is a change', () => {
+        spyOn(component, 'setupStream');
+        component.ngOnChanges();
+        expect(component.setupStream).toHaveBeenCalledTimes(1);
+    });
+
+    it('should setup canvas afterViewInit', () => {
+        component.ngAfterViewInit();
+        fixture.detectChanges();
+        console.log(canvasContext);
+        expect(canvasContext.fillStyle).toBe('#008000');
+    });
+
+    describe('fillMeter', () => {
+        beforeEach(() => {
+            spyOn(canvasContext, 'clearRect');
+            spyOn(canvasContext, 'fillRect');
+        });
+        it('should clear and fill canvas', () => {
+            component.fillMeter(1);
+
+            expect(canvasContext.clearRect).toHaveBeenCalledTimes(1);
```

Comment:
I don't think this assertion is required as you are already testing clearRect function is being called in the next line

Reply:
This ensures it's only be called once though. The next line could potentially check multiple calls if they were present.

Reply:
but what if in the future we do need to call this one more time because of some change in the requirement? Thats why I feel like we shouldn't test how many times a function is being called as that tests the code implementation and leave it open, rather a more appropriate test is what you have down here to check its being called with expected arguments. But if you think its necessary we can leave it.